lib/api/include/irods/getHostForGet.h

```diff
@@ -17,4 +17,9 @@ extern "C"
 #endif
 int rcGetHostForGet( rcComm_t *conn, dataObjInp_t *dataObjInp, char **outHost );
 
+#ifdef __cplusplus
+extern "C"
+#endif
+int rc_get_rescinfo_for_get( rcComm_t *conn, dataObjInp_t *dataObjInp, char **outHost );
```

Comment:
Does this make the original API obsolete?

Reply:
As in rcGetHostFor{Put,Get} being likely to be deprecated? Probably so.
What I haven't figured out is if it is still useful to return "THIS_ADDRESS" in the host string, as done by `rsGetHostForGet`.

Reply:
What does `THIS_ADDRESS` trigger?

Reply:
It is evidently a special value designed to prevent a `rcReconnect` in functino `redirectConnToRescSvr`. 
Who knows why....

Reply:
I think I'm against using THIS_ADDRESS in the new API.  It's a special string constant with a nebulous meaning.  And then what's to prevent someone from trying to use it on a different _comm object? That might have a different meaning , ie connect to a different machine.  Maybe better to stay with whatever `irods::get_loc_for_hier_string` places in the host variable here: https://github.com/irods/irods/blob/7e6fabaab4bb43ffa7e4b2a13e94a3b9ae7950af/server/api/src/rs_get_resource_info_for_operation.cpp#L72

Reply:
Think we've decided against THIS_ADDRESS being used, as it's weird or at best dated.  Returning even an empty hostname (if that's possible) is probably better. Will resolve convo tomorrow if no objections.

Reply:
`THIS_ADDRESS` is specific to the original `rcGetHostForPut/Get` functions and appears to be used in one location.

https://github.com/irods/irods/blob/f3ccaa842218e477395ebcf553639134433b63ee/lib/core/src/sockComm.cpp#L1618

What does your API return when you're already connected to the correct server?

Reply:
I believe it's more driven by the dataObjInp parameter (ie returns whatever hostname is preferred for that data object and the given operation) than whether the currently connected host is the correct one, which is why in the PRC we compare its return value to the currently connected host as input to the decision of whether to redirect.

Reply:
Thinking about his more, we probably need to return an indicator like `THIS_ADDRESS` instead of a hostname.

If we return a hostname, it's not guaranteed the client can resolve it.

We could include a boolean to indicate whether the client should continue to use the existing connection or connect to the server returned in the payload.
```json
{
    "redirect": false,
    "host": null,
    "resource_hierarchy": "pt;repl;ufs"
}
```

Thoughts?

Reply:
"redirect" would be a sign of whether an attempt should be made on resolving "host"; and the "host" value of `null` denotes the same meaning  "THIS_ADDRESS", correct ?

Maybe I have that wrong - I am not sure I'm understanding the purpose of "THIS_ADDRESS"

Reply:
Thinking about this more, perhaps returning the hostname of the connected server is fine because we have to assume any hostname returned is reachable by the client. _This needs to be documented._

The `redirect` property is a clear indicator to the client whether they need to open a new connection to the server identified by `host`.

Say a client calls this API endpoint and the server returns a hostname that identifies the one the client is already connected to, but the hostname is different (i.e. an alias). The `redirect` property comes in handy for situations like that.

Reply:
The names of the properties can be changed. I chose those to explain the approach.

Reply:
Yes, I really like the addition of the flag to indicate whether a redirect is necessary, just not sure how we calculate it.  Is that the condition that "THIS_ADDRESS" is actually signalling, and if so, is the iRODS server internally doing this via an IP- based equivalency-comparison between its computed hostname and the currently directly-connected host?

This is for my edification, as I still am a little in the dark : )

Reply:
Yes, it is acting as the concrete signal for `THIS_ADDRESS`.

I have to go review how the original APIs determine when to return `THIS_ADDRESS` to be sure.

Reply:
Turns out achieving the described behavior requires more work. `rxGetHostForGet` does not do anything smart to determine whether `THIS_ADDRESS` is returned. It just overwrites the value after hierarchy resolution is done.

`rxGetHostForPut` does not appear to set `THIS_ADDRESS` at all.

We need to think this through a little more.

Reply:
Given that this new API endpoint returns JSON, we can defer making it smarter to a later release.

For now, clients will just have to make an additional connection even if it's to the same server. What this means is clients shouldn't always call this API endpoint when transferring data. They need to use it for certain scenarios (e.g. large file transfers).

Thoughts?

Reply:
Right - that's fair - the client is responsible to consider when to use this functionality.   And we can make it better / add bumpers later.  Like you said, it's just JSON.

Reply:
Captured this conversation in #7336.

Reply:
@d-w-moore Feel free to mark this conversation as resolved.